pixeltable/functions/bfl.py

```diff
@@ -0,0 +1,512 @@
+                        retry_after_seconds = None
+                        retry_after_header = resp.headers.get('Retry-After')
+                        if retry_after_header is not None and re.fullmatch(r'\d{1,2}', retry_after_header):
+                            retry_after_seconds = int(retry_after_header)
+                        _logger.info(f'BFL task {task_id} polling rate limited, retry after: {retry_after_header}')
+                        raise BflRateLimitedError(
+                            f'BFL task {task_id} polling rate limited (429). retry-after:{retry_after_seconds}'
+                        )
+                    case _:
+                        error_text = await resp.text()
+                        _logger.info(f'BFL task {task_id} polling failed with status {resp.status}: {error_text}')
+                        raise BflUnexpectedError(f'BFL polling error (status {resp.status}): {error_text}')
+
+        raise BflUnexpectedError(f'BFL task {task_id} timed out after {max_wait} seconds')
+
+    async def _download_image(self, task_id: str, url: str) -> PIL.Image.Image:
+        """Download image from the result URL."""
+        async with self.session.get(url) as resp:
+            if resp.status != 200:
+                raise BflUnexpectedError(f'BFL task {task_id}: failed to download image, status {resp.status}')
+            img_data = await resp.read()
+            if len(img_data) == 0:
+                raise BflUnexpectedError(f'BFL task {task_id} resulted in an empty image')
+            img = PIL.Image.open(BytesIO(img_data))
+            img.load()
+            _logger.debug(
+                f'BFL task {task_id} successful. Image bytes: {len(img_data)}, size: {img.size}'
+                f', format: {img.format}, mode: {img.mode}'
+            )
+            return img
+
+    async def generate(self, endpoint: str, payload: dict) -> PIL.Image.Image:
+        """Submit a generation task and wait for the result."""
+        task_id, polling_url = await self._submit_task(endpoint, payload=payload)
+        _logger.debug(f'BFL task {task_id} submitted to {endpoint}')
+        return await self._poll_result(task_id, polling_url)
+
+
+@register_client('bfl')
+def _(api_key: str) -> _BflClient:
+    return _BflClient(api_key=api_key)
+
+
+def _client() -> _BflClient:
+    return Env.get().get_client('bfl')
+
+
+# Model endpoint mapping
+_MODEL_ENDPOINTS = {
+    # FLUX.2 models
+    'flux-2-pro': '/v1/flux-2-pro',
+    'flux-2-flex': '/v1/flux-2-flex',
+    'flux-2-max': '/v1/flux-2-max',
+    # FLUX 1.1 models
+    'flux-pro-1.1': '/v1/flux-pro-1.1',
+    'flux-pro-1.1-ultra': '/v1/flux-pro-1.1-ultra',
+    # FLUX.1 dev
+    'flux-dev': '/v1/flux-dev',
+    # Kontext models (for editing)
+    'flux-kontext-pro': '/v1/flux-kontext-pro',
+    'flux-kontext-max': '/v1/flux-kontext-max',
+}
+
+
+@pxt.udf(resource_pool='request-rate:bfl')
+async def generate(
+    prompt: str,
+    *,
+    model: Literal[
```

Comment:
Best to just take this as `model: str` rather than hardcode whatever models exist today.

Best not to specify a default for `model`.